sycl/source/detail/error_handling/enqueue_kernel.cpp

```diff
@@ -21,18 +22,40 @@ namespace detail {
 
 namespace enqueue_kernel_launch {
 
-bool oclHandleInvalidWorkGroupSize(const device_impl &DeviceImpl,
-                                   pi_kernel Kernel, const NDRDescT &NDRDesc) {
+bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
+                                const NDRDescT &NDRDesc) {
   const bool HasLocalSize = (NDRDesc.LocalSize[0] != 0);
 
   const plugin &Plugin = DeviceImpl.getPlugin();
   RT::PiDevice Device = DeviceImpl.getHandleRef();
+  cl::sycl::platform Platform = DeviceImpl.get_platform();
+
+  if (HasLocalSize) {
+    size_t MaxThreadsPerBlock[3] = {};
+    Plugin.call<PiApiKind::piDeviceGetInfo>(
+        Device, PI_DEVICE_INFO_MAX_WORK_ITEM_SIZES, sizeof(MaxThreadsPerBlock),
+        MaxThreadsPerBlock, nullptr);
+
+    for (size_t I = 0; I < 3; ++I) {
+      if (MaxThreadsPerBlock[I] < NDRDesc.LocalSize[I]) {
+        throw sycl::nd_range_error(
+            "The number of work-items in each dimension of a work-group cannot "
+            "exceed {" +
+                std::to_string(MaxThreadsPerBlock[0]) + ", " +
+                std::to_string(MaxThreadsPerBlock[1]) + ", " +
+                std::to_string(MaxThreadsPerBlock[2]) + "} for this device",
+            PI_INVALID_WORK_GROUP_SIZE);
+      }
+    }
+  }
 
   size_t VerSize = 0;
   Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION, 0,
                                           nullptr, &VerSize);
-  assert(VerSize >= 10 &&
-         "Unexpected device version string"); // strlen("OpenCL X.Y")
+  if (Platform.get_backend() == cl::sycl::backend::opencl) {
+    assert(VerSize >= 10 &&
+           "Unexpected device version string"); // strlen("OpenCL X.Y")
+  }
   string_class VerStr(VerSize, '\0');
   Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION,
                                           VerSize, &VerStr.front(), nullptr);
```

Comment:
This comment applies to the entire block of code:

```
  size_t VerSize = 0;
  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION, 0,
                                          nullptr, &VerSize);
  if (Platform.get_backend() == cl::sycl::backend::opencl) {
    assert(VerSize >= 10 &&
           "Unexpected device version string"); // strlen("OpenCL X.Y")
  }
  string_class VerStr(VerSize, '\0');
  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION,
                                          VerSize, &VerStr.front(), nullptr);
  const char *Ver = &VerStr[7]; // strlen("OpenCL ")
```

There are two problems here:

1. The `Ver` pointer could point to uninitialized memory when the backend is **not** OpenCL.
2. The name of the `Ver` variable should indicate that it is specific to OpenCL.

It would also be nice to include a comment, so readers understand why you are getting the version only for OpenCL. 
 I'd suggest rewriting the code to follow this form:

```
// Some of the error handling below is special for particular OpenCL versions.
// If this is an OpenCL backend, get the version.
const char *OpenClVer = nullptr;
string_class OpenClVerStr;
if (Platform.get_backend() == cl::sycl::backend::opencl) {
  size_t OclVerSize = 0;
  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION, 0,
                                          nullptr, &OclVerSize);
  assert(VerSize >= 10 &&
         "Unexpected device version string"); // strlen("OpenCL X.Y")
  OpenClVerStr.assign(OclVerSize, '\0');
  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_VERSION,
                                          OclVerSize, &OpenClVerStr.front(), nullptr);
  OpenClVer = &OpenClVerStr[7]; // strlen("OpenCL ")
}
```